File: sdks/typescript/trace-harness/src/view/tool-name.ts
const SCRIPT_SUFFIXES = new Set([
  ".py", ".sh", ".js", ".mjs", ".cjs", ".ts", ".tsx", ".rb", ".pl",
]);
const OPERATORS = new Set(["&&", "||", ";", "|"]);

function basename(value: string): string {
  return value.split(/[\\/]/).pop() ?? value;
}

function shellTokens(command: string): string[] | undefined {
  const tokens: string[] = [];
  let token = "";
  let quote: "'" | "\"" | undefined;
  const finish = () => {
    if (token) tokens.push(token);
    token = "";
  };
  for (let index = 0; index < command.length; index += 1) {
    const character = command[index]!;
    if (quote) {
      if (character === quote) quote = undefined;
      else if (character === "\\" && quote === "\"") {
        index += 1;
        if (index < command.length) token += command[index];
      } else token += character;
      continue;
    }
    if (character === "'" || character === "\"") {
      quote = character;
      continue;
    }
    if (character === "\\") {
      index += 1;
      if (index < command.length) token += command[index];
      continue;
    }
    if (/\s/.test(character)) {
      finish();
      continue;
    }
    const pair = command.slice(index, index + 2);
    if (pair === "&&" || pair === "||") {
      finish();
      tokens.push(pair);
      index += 1;
      continue;
    }
    if (character === ";" || character === "|") {
      finish();
      tokens.push(character);
      continue;
    }
    token += character;
  }
  if (quote) return undefined;
  finish();
  return tokens;
}

function commandName(command: unknown): string {
  if (typeof command !== "string" || !command.trim()) return "";
  const tokens = shellTokens(command) ?? command.trim().split(/\s+/);
  for (const token of tokens) {
    const candidate = token.replace(/[;&|]+$/, "");
    const dot = candidate.lastIndexOf(".");
    if (dot >= 0 && SCRIPT_SUFFIXES.has(candidate.slice(dot).toLowerCase())) {
      return basename(candidate);
    }
  }
  const segments: string[][] = [[]];
  for (const token of tokens) {
    if (OPERATORS.has(token)) segments.push([]);
    else segments.at(-1)!.push(token);
  }
  for (const segment of segments.reverse()) {
    const executable = segment.find((token) => (
      !token.startsWith("-") && !token.startsWith(">") && !token.startsWith("<")
      && !token.includes("=")
    ));
    if (executable && executable !== "cd") return basename(executable);
  }
  return "";
}

export function toolNameDetail(value: unknown): string {
  let argumentsValue = value;
  if (typeof argumentsValue === "string") {
    try {
      argumentsValue = JSON.parse(argumentsValue) as unknown;
    } catch {
      return "";
    }
  }
  if (!argumentsValue || typeof argumentsValue !== "object" || Array.isArray(argumentsValue)) {
    return "";
  }
  const argumentsRecord = argumentsValue as Record<string, unknown>;
  const command = commandName(argumentsRecord.command);
  if (command) return command;
  for (const key of ["file_path", "path", "filename", "file"]) {
    const candidate = argumentsRecord[key];
    if (typeof candidate === "string" && candidate) return basename(candidate);
  }
  return "";
}

File: sdks/python/trace_harness/view/tool_name.py
"""Tool-call names shared by Node Tree and AgentRun renderers."""

from __future__ import annotations

import json
import shlex
from pathlib import PurePath
from typing import Any

_SCRIPT_SUFFIXES = {
    ".py",
    ".sh",
    ".js",
    ".mjs",
    ".cjs",
    ".ts",
    ".tsx",
    ".rb",
    ".pl",
}


def _command_name(command: Any) -> str:
    """Prefer the invoked script over shell wrappers such as cd/python/redirection."""
    if not isinstance(command, str) or not command.strip():
        return ""
    try:
        tokens = shlex.split(command)
    except ValueError:
        tokens = command.split()

    for token in tokens:
        candidate = token.rstrip(";&|")
        if PurePath(candidate).suffix.lower() in _SCRIPT_SUFFIXES:
            return PurePath(candidate).name

    segments: list[list[str]] = [[]]
    for token in tokens:
        if token in {"&&", "||", ";", "|"}:
            segments.append([])
        else:
            segments[-1].append(token)
    for segment in reversed(segments):
        executable = next(
            (
                token
                for token in segment
                if not token.startswith(("-", ">", "<")) and "=" not in token
            ),
            "",
        )
        if executable and executable != "cd":
            return PurePath(executable).name
    return ""


def tool_name_detail(arguments: Any) -> str:
    """Extract the argument that best distinguishes repeated tool calls in a timeline."""
    if isinstance(arguments, str):
        try:
            arguments = json.loads(arguments)
        except (json.JSONDecodeError, TypeError):
            return ""
    if not isinstance(arguments, dict):
        return ""
    command = _command_name(arguments.get("command"))
    if command:
        return command
    for key in ("file_path", "path", "filename", "file"):
        value = arguments.get(key)
        if isinstance(value, str) and value:
            return PurePath(value).name
    return ""
